sycl/include/sycl/ext/oneapi/device_global/device_global.hpp

```diff
@@ -204,16 +201,80 @@ class
+  }                                                                            \
+                                                                               \
+  template <class RelayT = T>                                                  \
+  const std::remove_reference_t<                                               \
+      decltype(std::declval<RelayT>()[std::declval<std::ptrdiff_t>()])> &      \
+  operator[](std::ptrdiff_t idx) const noexcept {                              \
+    __SYCL_HOST_NOT_SUPPORTED("Subscript operator")                            \
+    return (*this->get_ptr())[idx];                                            \
+  }                                                                            \
+                                                                               \
+  template <class RelayT = T>                                                  \
+  std::enable_if_t<detail::HasArrowOperator<RelayT>::value ||                  \
+                       std::is_pointer_v<RelayT>,                              \
+                   RelayT> &                                                   \
+  operator->() noexcept {                                                      \
+    __SYCL_HOST_NOT_SUPPORTED("operator-> on a device_global")                 \
+    return *this->get_ptr();                                                   \
+  }                                                                            \
+                                                                               \
+  template <class RelayT = T>                                                  \
+  std::enable_if_t<detail::HasArrowOperator<RelayT>::value ||                  \
+                       std::is_pointer_v<RelayT>,                              \
+                   const RelayT> &                                             \
+  operator->() const noexcept {                                                \
+    __SYCL_HOST_NOT_SUPPORTED("operator-> on a device_global")                 \
+    return *this->get_ptr();                                                   \
+  }                                                                            \
+                                                                               \
+  template <typename propertyT> static constexpr bool has_property() {         \
+    return property_list_t::template has_property<propertyT>();                \
+  }                                                                            \
+                                                                               \
+  template <typename propertyT> static constexpr auto get_property() {         \
+    return property_list_t::template get_property<propertyT>();                \
+  }
+
 template <typename T, typename... Props>
 class
 #ifdef __SYCL_DEVICE_ONLY__
     [[__sycl_detail__::global_variable_allowed, __sycl_detail__::device_global,
```

Comment:
We could even do that with the current property. It should be as simple as doing a `std::conditional_t<property_list_t::template has_property<device_constant_key>(), __OPENCL_CONSTANT_AS__ T *, T *>`, inheriting the definition of `__OPENCL_CONSTANT_AS__` from sycl/include/sycl/access/access.hpp. Assuming the PTX and AMDGCN know how to handle `__attribute__((opencl_constant))`, that should hopefully avoid the need for the new property.

As for the member availability, this could be done through either conditionally picking base classes or SFINAE.

Reply:
> We could even do that with the current property. It should be as simple as doing a `std::conditional_t<property_list_t::template has_property<device_constant_key>(), __OPENCL_CONSTANT_AS__ T *, T *>`, inheriting the definition of `__OPENCL_CONSTANT_AS__` from sycl/include/sycl/access/access.hpp. Assuming the PTX and AMDGCN know how to handle `__attribute__((opencl_constant))`, that should hopefully avoid the need for the new clang attribute.
> 
> As for the member availability, this could be done through either conditionally picking base classes or SFINAE.

Doing address space declarations directly in source code is currently not allowed by SEMA: I get e.g. (same if field is pointer type):
```
error: field may not be qualified with an address space
  100 |   T __attribute__((opencl_constant)) val{};
```

I looked into changing this behaviour, but I didn't think there was a simple solution.
My current idea is to partially specialize as
```
    device_global<
        T, detail::properties_t<Props...>,
        typename std::enable_if_t<(detail::properties_t<Props...>::
                               template has_property<device_constant_key>())>>
    : public detail::device_global_base<T, detail::properties_t<Props...>>
```
such that when the property device_constant is used we add a clang attribute to the class :

``__sycl_detail__::device_constant``

That the compiler then uses to manually set the address space to .const *only* for cuda/hip backends.

This I think in theory should be compatible with the partial specializations of the `device_global_base` class; to allow simultaneous specializations for the case when `device_image_scope` property is in the property list (via device_global_base): The only property that can be used in combination with the `device_constant` property should be `device_image_scope` I think (I will update the specification doc once the implementation is finalised).

 @steffenlarsen @Naghasan Maybe there is a better solution though?

Once this is done there should also probably be a few more tests added to check all valid combined functionality of the device_constant property with the device_image_scope property. There should also be a test checking that the compiler does not allow writing to a device_global with the device_constant property (apart from via sycl::hander::/queue::memcpy etc).

Reply:
If we need the attribute, could we maybe make it apply to the field instead of the device-global class then? I.e. that way we can use the conditional solution we previously discussed, but with the new attribute.

As a thought experiment, can we think of a case where a const global variable (including const fields of non-const global variables) would not want the variables to be in the `.const` namespace for NVPTX and AMDGCN? If not, could we maybe make the address-space decision based on that?

Reply:
> If we need the attribute, could we maybe make it apply to the field instead of the device-global class then? I.e. that way we can use the conditional solution we previously discussed, but with the new attribute.
> 
Yeah this sounds like it might be a better solution. I'll look into doing this.

> As a thought experiment, can we think of a case where a const global variable (including const fields of non-const global variables) would not want the variables to be in the `.const` namespace for NVPTX and AMDGCN? If not, could we maybe make the address-space decision based on that?

I am not sure if it can be an issue for NVPTX and AMDGCN backends, but in theory you can run out of .const memory space (It is normally limited to 4kb), and hence a user may want to be careful about which variables to put in .const space. If it is not a real issue in these backends, it could be an issue in other as yet unsupported backends. This is why we do not wan to have `device_global<const T>` imply .const address space, and instead have the explicit `device_global<T, decltype(properties{device_constant})>` solution, which will be functionally identical to ``device_global<const T, decltype(properties{device_constant})>``:

`device_constant` implies `const T` but `const T` shouldn't imply .const address space.